### DISCUSSION TITLE
%%sparql --display-properties

### DISCUSSION THREAD

**Discussion by CoryCasanave** [aws/graph-notebook#647] [Q&A] [answered]
I am using %%sparql to show a business oriented graph display. The nodes and edges should use names and labels, not URI. (note the names are not unique). I have tried every variation of --display-property I can think of and nothing works. Please post an example of this feature. The construct portion of my query is as follows:
```
construct {
  ?node :directParent ?directParent .
  ?node :ultimateParent ?ultimateParent .
  ?node :name ?name .
}
```

**Comment by michaelnchin** [comment, marked as answer]
Hi @CoryCasanave , thank you for reaching out! 

The effectiveness of the `-d`/`--display-property` and `-de`/`--edge-display-property` options for `%%sparql` can vary, depending on how much information is returned for each individual element in the query results. In this respect, visualized vertexes created from the `subject` of a returned binding are typically more flexible than `object` vertexes.

To demonstrate, let's use bindings returned by a construct query from the [EPL-SPARQL](https://github.com/aws/graph-notebook/blob/main/src/graph_notebook/notebooks/02-Visualization/EPL-SPARQL.ipynb) sample notebook:
```
%%sparql -l 100 -le 100

PREFIX loco: <http://www.example.com/location/ontology/>
PREFIX locr: <http://www.example.com/location/resource#>
PREFIX soco: <http://www.example.com/soccer/ontology/>
PREFIX socr: <http://www.example.com/soccer/resource#>
prefix so:   <https://schema.org/>
        
CONSTRUCT {?s ?p ?o}
WHERE {
  ?s ?p ?o .
  ?s a soco:Team 
} 
```

**For vertexes representing an `object` of type URI in a binding:**

Example binding:
```
{
        "subject": {
          "type": "uri",
          "value": "http://www.example.com/soccer/resource#Leicester_City"
        },
        "predicate": {
          "type": "uri",
          "value": "http://www.example.com/soccer/ontology/homeStadium"
        },
        "object": {
          "type": "uri",
          "value": "http://www.example.com/soccer/resource#King_Power_Stadium"
        }
}
```

For `object` vertexes, the only information available to the graph visualizer is within the binding itself - namely, the `type` and `value` key/value pairs. 

By default, the visualizer will use the URI given by `value` as the displayed label. Additionally, if a matching `PREFIX` has been specified with the query, this prefix will also be substituted into the full URI. 

For example, with the above object URI:`http://www.example.com/soccer/resource#King_Power_Stadium`, and the provided prefix `socr: <http://www.example.com/soccer/resource#>`, the final label will be displayed as `socr:King_Power_Stadium`.

If you would like to set the `type` value as the displayed label instead of the default, you can run `%%sparql -d type`. To display the entire URI from the `value` field, use `%%sparql -d value`.

**For vertexes representing a`subject` in a binding:**

`subject` vertexes are the most flexible, as they can display the value of any literal property that they are linked to via a S/P/O binding triple. 

By default, the graph visualizer will always try to find and use the value of the `...#label` property. For example, if the following label binding is present for the `...#Leicester_City` subject node:

```
{
        "subject": {
          "type": "uri",
          "value": "http://www.example.com/soccer/resource#Leicester_City"
        },
        "predicate": {
          "type": "uri",
          "value": "http://www.w3.org/2000/01/rdf-schema#label"
        },
        "object": {
          "type": "literal",
          "value": "Leicester City"
        }
}
```

Then the visualizer will set the default displayed label as `Leicester City`.

If you would like to set the subject vertex's displayed label to a different literal property's value, you will need to prepend the `P.` prefix to the property name (AKA the predicate value). The name can be either the full URI, or a shortened URI with any prefix substitutions indicated in the original query. 

For example, with the following binding for literal property `.../nickname`:
```
{
        "subject": {
          "type": "uri",
          "value": "http://www.example.com/soccer/resource#Leicester_City"
        },
        "predicate": {
          "type": "uri",
          "value": "http://www.example.com/soccer/ontology/nickname"
        },
        "object": {
          "type": "literal",
          "value": "The Foxes"
        }
}
```

The full predicate value is `http://www.example.com/soccer/ontology/nickname`, which can shortened using the `soco` prefix to become `soco:nickname`. The query to label all eligible nodes with their nicknames would be:

```
%%sparql -d P.soco:nickname
...
```

Let me know if this helped, and of any additional questions!
